### PR TITLE
Disable implicit casts

### PR DESCRIPTION
**Pull request by isoos**
Seems like we don't have any potential bug, but nevertheless...

### CODE REVIEW

**Comment by jakobr-google** [comment]
Another good thing to check is to run the tests in Dart 2 mode. I've been using
```
find test -name \*_test.dart -exec dart --preview-dart-2 {} \;
```
which catches a bunch of these, and possibly more.

I've been fixing Dart 2 issues in the libraries that Pub uses, so we should also make sure to upgrade to the latest versions of appengine, gcloud, googleapis, and memcache to get the Dart 2 runtime fixes.

File: app/bin/service/search.dart [hunk 1]
```diff
@@ -78,7 +78,7 @@ void _main(FrontendEntryMessage message) {
     final scheduler = new TaskScheduler(
       batchIndexUpdater,
       [
-        new ManualTriggerTaskSource(taskReceivePort),
+        new ManualTriggerTaskSource(taskReceivePort as Stream<Task>),
```
Comment (jakobr-google) [inline app/bin/service/search.dart]: Are you sure this works in Dart 2 mode? If I'm reading things right, a `ReceivePort` is a `Stream<dynamic>`, so this cast should fail at runtime.

File: app/lib/frontend/backend.dart [hunk 2]
```diff
@@ -191,7 +198,7 @@ class GCloudPackageRepository extends PackageRepository {
         onPause: () => subscription.pause(),
         onResume: () => subscription.resume(),
         onCancel: () => subscription.cancel());
-    return controller.stream;
+    return controller.stream.cast<PackageVersion>();
```
Comment (jakobr-google) [inline app/lib/frontend/backend.dart]: Instead, you could make `controller` a `StreamController<PackageVersion>`:
```dart
  final controller = new StreamController<PackageVersion>(...);
```
Then you wouldn't need to cast here.

File: app/lib/frontend/backend.dart [hunk 3]
```diff
@@ -126,27 +130,30 @@ class Backend {
     final packageVersionKey = db.emptyKey
         .append(models.Package, id: package)
         .append(models.PackageVersion, id: version);
-    return (await db.lookup([packageVersionKey])).first;
+    return (await db.lookup([packageVersionKey]))
+        .cast<models.PackageVersion>()
```
Comment (jakobr-google) [inline app/lib/frontend/backend.dart]: Using `cast()` wraps the list in a new object that does the cast on every access. It doesn't matter so much here, but I believe the recommendation is to cast the object returned by `first` instead:
```dart
return (await db.lookup(...)).first as PackageVersion;
```
https://www.dartlang.org/guides/language/effective-dart/usage#avoid-using-cast

**Comment by isoos** [comment]
Fixed most things, but lot of the dependent libraries are not dart2-ready yet.

**Comment by jakobr-google** [comment]
I've done a lot of Dart 2 readiness updates to package:appengine, gcloud, googleapis, and memcache recently, so make sure you get the latest available versions of those.

**Comment by isoos** [comment]
Yeah, thanks for noting that. I've noticed that we have a few potential upgrades, but it wasn't picked up yet, will take a look.

**Pull request merged**